### ISSUE TITLE
Hanging thread on dynamic imports using cx_Freeze 5.0.1

### ISSUE THREAD

**Issue by dandersson** [marcelotduarte/cx_Freeze#270]
Using

* cx_Freeze 5.0.1 (does _not_ happen on 4.3.4)
* Tornado 4.5.1 (4.1 behaves similarly)
* Python 2.7.13 32-bit
* Windows 7 64-bit

While trying to upgrade to cx_Freeze 5.0.1 from cx_Freeze 4.3.4, I found issues regarding threading and specifically Tornado, though it is probably a more generic issue. It might be very related to https://github.com/anthony-tuininga/cx_Freeze/issues/269, but I will open a separate issue to begin with.

In short, my application starts a Tornado web server in a separate thread. When running with cx_Freeze 5.0.1, that thread hangs when starting the IO loop, though it does not happen in non-frozen mode, nor with cx_Freeze 4.3.0.

I have boiled it down to a smaller example that exhibits this behaviour (with hefty logging included to make sense of the threading):

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import logging
import threading
import time

import tornado.ioloop
import tornado.netutil
import tornado.web


class MainHandler(tornado.web.RequestHandler):
    def get(self):
        self.write('Hello, world')


def make_app():
    return tornado.web.Application([(r"/", MainHandler),])


def run_web_server():
    logger = logging.getLogger('run_web_server')
    logger.info('Running web server')

    logger.debug('Making app')
    app = make_app()
    logger.debug('Getting sockets')
    sockets = tornado.netutil.bind_sockets(0)
    logger.debug('Creating HTTPServer')
    server = tornado.httpserver.HTTPServer(app)
    logger.debug('Adding sockets to server')
    server.add_sockets(sockets)
    logger.debug('Sockets added')

    for socket in sockets:
        logger.debug('Listening on %s, port %d' % socket.getsockname()[:2])

    logger.debug('Starting IOLoop')
    tornado.ioloop.IOLoop.current().start()
    logger.debug('IOLoop done')


def counter(n=5):
    logger = logging.getLogger('counter')
    for i in range(n):
        logger.info('Counting %d', i + 1)
        time.sleep(1)


def main():
    logging.basicConfig(
        level=logging.DEBUG,
        format='%(asctime)s::%(threadName)-10s::%(levelname)-5s::%(funcName)15s::%(name)-15s:: %(message)s'
    )
    logger = logging.getLogger('main')

    # tornado.ioloop.IOLoop.current()
    web_server_thread = threading.Thread(target=run_web_server)

    try:
        web_server_thread.start()
        counter()
    finally:
        tornado.ioloop.IOLoop.instance().stop()


if __name__ == '__main__':
    main()
```

Running this code through a Python interpreter will start the Tornado IO loop in its separate thread accepting requests on `localhost`, count to 5 in the main thread and then exit cleanly.

Running the same code after it having been frozen using cx_Freeze 5.0.1 will make the Tornado thread hang at the line below the "Starting IOLoop" log message. The counter will run, but the Tornado IO loop never starts.

Chasing through the Tornado code, it hangs while having obtained a `threading.Lock` when it tries to return the IO loop implementation in the `tornado.ioloop.IOLoop.configurable_default` method, at the line

```python
        from tornado.platform.select import SelectIOLoop
```

It seems like this import when running in a thread context just hangs, which means that the earlier mentioned lock is not released.

Now, note the commented line in the `main` function. If it is uncommented, the thread will _not_ hang. Tornado in general keeps track of a singleton `IOLoop` instance per process. The effect that accessing this singleton IO loop object early has is to create it early from the main thread, instead of creating it on first access in the web server thread.

We can note the [warning message regarding imports in the Python `threading` documentation](https://docs.python.org/2/library/threading.html#importing-in-threaded-code):

> While the import machinery is thread-safe, there are two key restrictions on threaded imports due to inherent limitations in the way that thread-safety is provided:
> 
> * Firstly, other than in the main module, an import should not have the side effect of spawning a new thread and then waiting for that thread in any way. Failing to abide by this restriction can lead to a deadlock if the spawned thread directly or indirectly attempts to import a module.
> * Secondly, all import attempts must be completed before the interpreter starts shutting itself down. This can be most easily achieved by only performing imports from non-daemon threads created through the threading module. Daemon threads and threads created directly with the thread module will require some other form of synchronization to ensure they do not attempt imports after system shutdown has commenced. Failure to abide by this restriction will lead to intermittent exceptions and crashes during interpreter shutdown (as the late imports attempt to access machinery which is no longer in a valid state).

I am not sure if this is related in this case. In any case, it should not differ between running a Python interpreter directly (or through cx_Freeze 4.3.4) or running it after freezing with cx_Freeze 5.0.1.

I froze the application using:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import cx_Freeze


cx_Freeze.setup(executables=[cx_Freeze.Executable('thread_test.py')])
```

`requirements.txt`:

```
cx-Freeze==5.0.1
tornado==4.5.1
```

I also have to apply the [workaround mentioned in issue 209](https://github.com/anthony-tuininga/cx_Freeze/issues/209#issuecomment-290673644), moving everything in `build/exe.win32-2.7/libs` to `build/exe.win32-2.7` to  to be able to start the application in frozen mode using cx_Freeze 5.0.1 (but not when using cx_Freeze 4.3.4).

**Comment by anthony-tuininga** [comment]
Does the patch already applied (https://github.com/anthony-tuininga/cx_Freeze/pull/245) resolve this for you? I plan to make a 5.0.2 sometime in the next few weeks -- but it would be good to know if this resolves the matter for you!

**Comment by dandersson** [comment]
@anthony-tuininga It certainly sounds related! If I can find the time I will test it tomorrow, otherwise early next week. Thanks for the quick response!

**Comment by anthony-tuininga** [comment]
No problem. :+1:

**Comment by dandersson** [comment]
@anthony-tuininga I can verify that #245 indeed fixed the issue! I will run this locally built version of cx-Freeze until 5.0.2 is released.

**Comment by anthony-tuininga** [comment]
That's great. Thanks for confirming that.

**Comment by gricey432** [comment]
I've been running into this for the last few days in a Flask app, seems to be fixed now in 5.0.2. Thanks @anthony-tuininga
